### PR TITLE
feat: integrate `@cryptkeeperzk/semaphore-proof` package

### LINKED ISSUE #590: fix: remove `ZkProofService.Semaphore` from provider
Shouldn't generate Semaphore proof from the injected provider.

### LINKED ISSUE #625: fix: switch to forked `@semaphore-protocol/proof`

### PR DESCRIPTION
**Pull request by 0xisk**
## Explanation

This PR is for generating Semaphore proofs on Chrome browser platforms through using `Offscreen` feature, that's because of known bug issue in MV3 https://bugs.chromium.org/p/chromium/issues/detail?id=1219164
 
## More Information

* closes #590 
* closes #625 

## Screenshots/Screencaps

N/A

## Manual Testing Steps

1. Build extension using `pnpm run dev`
2. Run MockMerkle server `pnpm run merkle:start`
3. Run Demo `pnpm run demo:start`
4. Connect to the demo.
5. Generate a SemaphoreProof from demo. 
6. Expected to see the FullProof output in the demo page. 

Screenshot_20230714_155039

## Pre-Merge Checklist

- [X] PR template is filled out
- [X] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [x] PR has been added to the appropriate release Milestone

> PR template source from [github.com/MetaMask](https://github.com/MetaMask)

### CODE REVIEW

File: packages/app/src/background/services/injector/__tests__/injector.test.ts [hunk 1]
```diff
@@ -45,8 +46,11 @@ jest.mock("@src/background/services/zkIdentity", (): unknown => ({
   })),
 }));
 
+jest.mock("@src/util/pushMessage");
+
 describe("background/services/injector", () => {
   beforeEach(() => {
+    (pushMessage as jest.Mock).mockReset();
```
Comment (0xmad) [inline packages/app/src/background/services/injector/__tests__/injector.test.ts:53]:
```suggestion
    (pushMessage as jest.Mock).mockClear();
```

File: packages/providers/src/sdk/CryptKeeperInjectedProvider.ts [hunk 2]
```diff
@@ -320,32 +319,28 @@ export class CryptKeeperInjectedProvider {
...
    * @returns {Promise<SemaphoreProof>} A Promise that resolves to the semaphore proof.
    */
-  async semaphoreProof(
+  async generateSemaphoreProof(
     externalNullifier: string,
     signal: string,
     merkleProofArtifactsOrStorageAddress: string | MerkleProofArtifacts,
     merkleProof?: MerkleProof,
-  ): Promise<SemaphoreProof> {
+  ): Promise<SemaphoreProof | void> {
```
Comment (0xmad) [inline packages/providers/src/sdk/CryptKeeperInjectedProvider.ts:327]: Why is it union with void?

File: packages/app/src/util/pushMessage.ts [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
 
-import { ReduxAction } from "@src/types";
+import { ReduxAction, RequestHandler } from "@src/types";
 
-export default async function pushMessage(message: ReduxAction): Promise<void> {
+export default async function pushMessage(message: ReduxAction | RequestHandler): Promise<unknown> {
   try {
-    await browser.runtime.sendMessage(message);
+    const result = (await browser.runtime.sendMessage(message)) as unknown;
+    return result;
```
Comment (0xmad) [inline packages/app/src/util/pushMessage.ts:9]:
```suggestion
    const result = await browser.runtime.sendMessage(message);
    return result as unknown;
```

File: packages/app/src/offscreen/index.ts [hunk 4]
```diff
@@ -1,3 +1,34 @@
...
+  }
+
+  try {
+    const offscreenController = new OffscreenController();
+    offscreenController.initialize();
+
+    const result = await offscreenController.handle(request, sender);
+    return result;
+  } catch (e) {
+    throw new Error("Proof gneeration is failed");
```
Comment (0xmad) [inline packages/app/src/offscreen/index.ts:24]:
```suggestion
    throw new Error("Proof generation is failed");
```

File: packages/app/src/offscreen/index.ts [hunk 5]
```diff
@@ -1,3 +1,34 @@
...
+  } finally {
+    browser.runtime.onMessage.removeListener(offscreenMessageListener);
+  }
+};
+
+try {
+  browser.runtime.onMessage.addListener(offscreenMessageListener);
```
Comment (0xmad) [inline packages/app/src/offscreen/index.ts:31]: Does this function throws error?

File: packages/app/src/offscreen/__tests__/offscreenScript.test.ts
Comment (0xmad) [inline packages/app/src/offscreen/__tests__/offscreenScript.test.ts:1]: Let's remove this file.

File: packages/app/src/offscreen/__tests__/offscreenController.test.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+  beforeEach(() => {
+    (ZkIdentitySemaphore.genFromSerialized as jest.Mock).mockReturnValue("serialized");
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  // test("should be able to initialize Offscreen controller", async () => {
```
Comment (0xmad) [inline packages/app/src/offscreen/__tests__/offscreenController.test.ts:39]: Let's remove it

File: packages/app/src/background/services/injector/index.ts [hunk 7]
```diff
@@ -65,6 +69,77 @@ export default class InjectorService {
...
+    if (!permission.canSkipApprove) {
+      await this.requestManager.newRequest(PendingRequestType.SEMAPHORE_PROOF, {
+        ...semaphoreRequest,
+        origin: meta.origin,
+      });
+    }
+
+    try {
+      if (browserPlatform !== BrowserPlatform.Firefox) {
+        return (await pushMessage({
```
Comment (0xmad) [inline packages/app/src/background/services/injector/index.ts:125]: It should be a separate variable

File: packages/app/src/background/services/injector/index.ts [hunk 8]
```diff
@@ -65,6 +69,77 @@ export default class InjectorService {
     }
   };
 
+  // TODO: writing tests
+  generateSemaphoreProof = async (
+    { merkleStorageAddress, externalNullifier, signal, merkleProofArtifacts, merkleProof }: SemaphoreProofRequest,
+    meta: IMeta,
+  ): Promise<SemaphoreProof | undefined> => {
```
Comment (0xmad) [inline packages/app/src/background/services/injector/index.ts:76]:
```suggestion
  ): Promise<SemaphoreProof> => {
```

File: packages/app/src/background/services/injector/index.ts [hunk 9]
```diff
@@ -65,6 +69,77 @@ export default class InjectorService {
...
+      });
+    }
+
+    try {
+      if (browserPlatform !== BrowserPlatform.Firefox) {
+        return (await pushMessage({
+          method: RPCAction.GENERATE_SEMAPHORE_PROOF_OFFSCREEN,
+          payload: {
+            ...semaphoreRequest,
+          },
```
Comment (0xmad) [inline packages/app/src/background/services/injector/index.ts:129]:
```suggestion
          payload: semaphoreRequest,
```

File: packages/app/src/offscreen/Offscreen.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      identityCommitment,
+      merkleProofArtifacts,
+    });
+
+    const fullProof = await generateProof(identityGenerated.zkIdentity, merkleProof, externalNullifier, signal, {
+      wasmFilePath: circuitFilePath,
+      zkeyFilePath,
+    });
+
+    return fullProof;
```
Comment (0xmad) [inline packages/app/src/offscreen/Offscreen.ts:52]:
```suggestion
    return generateProof(identityGenerated.zkIdentity, merkleProof, externalNullifier, signal, {
      wasmFilePath: circuitFilePath,
      zkeyFilePath,
    });
```

File: packages/app/src/offscreen/Offscreen.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+import Handler from "@src/background/controllers/handler";
+import { Runtime } from "webextension-polyfill";
+
+const RPC_METHOD_ACCESS: Record<RPCAction, boolean> = {
+  ...Object.values(RPCAction).reduce((acc, method) => ({ ...acc, [method]: false }), {} as Record<RPCAction, boolean>),
+  [RPCAction.GENERATE_SEMAPHORE_PROOF_OFFSCREEN]: true,
+};
+
+export class OffscreenController {
```
Comment (0xmad) [inline packages/app/src/offscreen/Offscreen.ts:14]: Why do you need another handler?

File: packages/app/src/offscreen/__tests__/offscreenController.test.ts (hunk 6 above)
Comment (0xisk) [reply packages/app/src/offscreen/__tests__/offscreenController.test.ts:39]: That's for another PR.

File: packages/app/src/offscreen/Offscreen.ts (hunk 11 above)
Comment (0xisk) [reply packages/app/src/offscreen/Offscreen.ts:14]: It's needed to handle Offscreen RPC related calls.

File: packages/app/src/background/services/injector/index.ts [hunk 12]
```diff
@@ -65,6 +69,79 @@ export default class InjectorService {
...
+        const fullProof =  (await pushMessage({
+          method: RPCAction.GENERATE_SEMAPHORE_PROOF_OFFSCREEN,
+          payload: {
+            ...semaphoreRequest,
+          },
+          meta,
+          source: "offscreen",
+        })) as Promise<SemaphoreProof>;
+
+        return fullProof;
```
Comment (0xmad) [inline packages/app/src/background/services/injector/index.ts:134]:
```suggestion
        const fullProof =  await pushMessage({
          method: RPCAction.GENERATE_SEMAPHORE_PROOF_OFFSCREEN,
          payload: semaphoreRequest,
          meta,
          source: "offscreen",
        });

        return fullProof as Promise<SemaphoreProof>;
```

**Review by 0xmad — APPROVED** [review]
@0xisk thanks, just need to pass the checks.

**Pull request merged**